### PR TITLE
Xianya/walmart

### PR DESCRIPTION
**Pull request by VictorYXL**
# Description

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.7
  - [ ] 3.8
  - [ ] 3.9
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 1]
```diff
@@ -10,41 +10,75 @@
...
-        self.in_transit_snapshot: Dict[int, Dict[int, int]] = defaultdict(dict)
+        # Use tuple(facility_name, sku_name) as index
+        self.product_level_snapshot: Dict[tuple(str, str), Dict[int, int]] = defaultdict(dict)
+        self.in_transit_snapshot: Dict[tuple(str, str), Dict[int, int]] = defaultdict(dict)
+        self.stock_quantity: Dict[tuple(str, str), Dict[int, int]] = defaultdict(dict)
+
+    def load_data(self, state: dict, history_start: int) -> pd.DataFrame:
+        cost = state["upstream_price_mean"]
+        # Load history and today data from env
+        target_df = pd.DataFrame(columns=["Price", "Cost", "Demand"])
```
Comment (VictorYXL) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:39]: Demand or Cost?
是否有必要允许自定义列名(oracle file)？

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            if "preprocessed" not in oracle_name:
+                oracle_file_path = os.path.join(oracle_file_dir, oracle_name)
+                if oracle_file_path.endswith(".csv"):
+                    df_list.append(pd.read_csv(oracle_file_path, parse_dates=["Date"]))
+                elif oracle_file_path.endswith(".tsv"):
+                    df_list.append(pd.read_csv(oracle_file_path, parse_dates=["Date"], sep="\t"))
+                elif oracle_file_path.endswith(".xlsx"):
+                    df_list.append(pd.read_excel(oracle_file_path, parse_dates=["Date"]))
+                else:
+                    raise NotImplementedError
```
Comment (VictorYXL) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:30]: 报错还是跳过？

File: examples/supply_chain/rl/config.py [hunk 3]
```diff
@@ -84,11 +86,15 @@ def get_vlt_buffer_factor(entity: SupplyChainEntity, facility_info: FacilityInfo
...
-    "data_loader": "DataLoaderFromFile",
-    "oracle_file": "oracle_samples.csv",  # Only need in DataLoaderFromFile loader
-    "history_len": 28,  # E.g., mapping to np.inf in instance creation if it is static
-    "future_len": 7,
-    "update_frequency": 7,  # E.g., mapping to np.inf in instance creation if no update
+    "forecastor": "OracleForecastor",
+    # Oracle file, start_date_time and durations are only needed in OracleForecastor
+    # All files with suffix csv, tsv and xlsx without preprocessed will be loaded.
+    "oracle_file_dir": "maro/simulator/scenarios/supply_chain/topologies/walmart_3_layers/data",
+    "start_date_time": "2021-07-01",
```
Comment (VictorYXL) [inline examples/supply_chain/rl/config.py:93]: 能否从config.yml中获取？

File: examples/supply_chain/rl/forecastor/base_forecastor.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+# Licensed under the MIT license.
+
+import pandas as pd
+
+
+class BaseForecastor(object):
+    def __init__(self, data_loader_conf) -> None:
+        self.data_loader_conf = data_loader_conf
+
+    def forecast_future_demand(self, state: dict, history_df: pd.DataFrame) -> pd.DataFrame:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/base_forecastor.py:11]: rename to `df_history`

File: examples/supply_chain/rl/forecastor/moving_average_forecastor.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import pandas as pd
+
+from .base_forecastor import BaseForecastor
+
+
+class MovingAverageForecastor(BaseForecastor):
+    def forecast_future_demand(self, state: dict, history_df) -> pd.DataFrame:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/moving_average_forecastor.py:10]: add type hint

File: examples/supply_chain/rl/forecastor/moving_average_forecastor.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+        his_demand_mean = history_df["Demand"].mean().item()
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/moving_average_forecastor.py:11]: `history`

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import datetime
+import os
+
+import pandas as pd
+
+from .base_forecastor import BaseForecastor
+
+
+class OracleForecastor(BaseForecastor):
+    def __init__(self, data_loader_conf) -> None:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:13]: type hint

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        super().__init__(data_loader_conf)
+
+        oracle_file_dir = self.data_loader_conf["oracle_file_dir"]
+        df_list = []
+        for oracle_name in os.listdir(oracle_file_dir):
+            # The simulator will product the preprocessed file in the target dir.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:19]: add "TODO" mark

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            # To avoid the confilct, currently the file name with preprocessed will be skipped.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:20]: typo: conflict

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py (hunk 8 above)
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:19]: typo: produce?

Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:19]: Extension: Code Spell Checker

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py (hunk 2 above)
Comment (Jinyu-W) [reply examples/supply_chain/rl/forecastor/oracle_forecastor.py:30]: decide later, TODO :)

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py (hunk 7 above)
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:13]: Add docstring to describe the required headers of the oracle file:
```
Date:
sku_name:
facility_name:
....
```

Comment (Jinyu-W) [reply examples/supply_chain/rl/forecastor/oracle_forecastor.py:13]: 给这个类加docstring，说明一下要求的oracle文件的各列列名及含义

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                elif oracle_file_path.endswith(".xlsx"):
+                    df_list.append(pd.read_excel(oracle_file_path, parse_dates=["Date"]))
+                else:
+                    raise NotImplementedError
+        self.df_raws = pd.concat(df_list, axis=0)
+        self.df_raws.sort_values(by="Date", ascending=True)
+        self.start_date = datetime.datetime.strptime(self.data_loader_conf["start_date_time"], "%Y-%m-%d")
+        self.date_len = self.data_loader_conf["durations"]
+
+    def forecast_future_demand(self, state: dict, history_df: pd.DataFrame) -> pd.DataFrame:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:36]: add docstring to describe the required state keys

File: maro/rl/policy/abs_policy.py [hunk 11]
```diff
@@ -96,9 +96,17 @@ def train(self) -> None:
...
 
-    def __init__(self, name: str) -> None:
+    Args:
+        name(str): name of this policy
+        action_by_idx (bool, default=True):
+            When set to true, the action quantity will be set as action index,
+            or the action will be determined by quantity.
+    """
+
+    def __init__(self, name: str, action_by_idx: bool = True) -> None:
```
Comment (Jinyu-W) [inline maro/rl/policy/abs_policy.py:108]: Do not add the arg for this base policy.
Implement it inside `examples/supply_chain`

File: examples/supply_chain/rl/env_sampler.py [hunk 12]
```diff
@@ -301,6 +301,16 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
         else:
             return 0.0
 
+    def _get_upstream_sku_info(self, facility_id: int, sku_id: int) -> Optional[Dict[int, SkuInfo]]:
+        upstreams_sku_ids = self._facility_info_dict[facility_id].upstreams
+        if sku_id in upstreams_sku_ids:
+            upstream_facility_ids = upstreams_sku_ids[sku_id]
+            facility_list = [self._facility_info_dict[facility_id] for facility_id in upstream_facility_ids]
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:308]: rename to `facility_info_list`

File: examples/supply_chain/rl/env_sampler.py [hunk 13]
```diff
@@ -301,6 +301,16 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
...
+            facility_id_to_sku_info = {facility.id: facility.skus[sku_id] for facility in facility_list}
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:309]: rename to `facility_info`

File: examples/supply_chain/rl/env_sampler.py [hunk 14]
```diff
@@ -301,6 +301,16 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
         else:
             return 0.0
 
+    def _get_upstream_sku_info(self, facility_id: int, sku_id: int) -> Optional[Dict[int, SkuInfo]]:
+        upstreams_sku_ids = self._facility_info_dict[facility_id].upstreams
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:305]:
```
sku_id_to_facility_id_list: Dict[int, List[int]] = ...
```

File: examples/supply_chain/rl/env_sampler.py [hunk 15]
```diff
@@ -301,6 +301,16 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
         else:
             return 0.0
 
+    def _get_upstream_sku_info(self, facility_id: int, sku_id: int) -> Optional[Dict[int, SkuInfo]]:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:304]: rename

File: examples/supply_chain/rl/env_sampler.py [hunk 16]
```diff
@@ -379,6 +389,13 @@ def get_or_policy_state(self, entity: SupplyChainEntity) -> dict:
         if self._storage_capacity_dict is None:
             self._storage_capacity_dict = self._get_storage_capacity_dict_info()
 
+        upstream_skus = self._get_upstream_sku_info(entity.facility_id, entity.skus.id)
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:392]: add type hint

File: examples/supply_chain/rl/algorithms/base_stock_policy.py [hunk 17]
```diff
@@ -85,43 +119,45 @@ def calculate_stock_quantity(
...
+            for i in range(min(self.history_start, state["cur_vlt"])):
+                purchased_before_action[-i - 1] = state["history_purchased"][self.history_start - i - 1]
+            self.stock_quantity[index] = self.calculate_stock_quantity(
                 target_df,
-                self.product_level_snapshot[entity_id][self.history_start],
-                self.in_transit_snapshot[entity_id][self.history_start],
-                state["cur_vlt"],
+                self.product_level_snapshot[index].get(self.history_start, 0),
+                self.in_transit_snapshot[index].get(self.history_start, 0),
+                # Since maro is performing the action at the end of each day, vlt needs to be deduced by 1
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/algorithms/base_stock_policy.py:145]: decrease

File: examples/supply_chain/rl/config.py [hunk 18]
```diff
@@ -45,7 +47,7 @@ def get_vlt_buffer_factor(entity: SupplyChainEntity, facility_info: FacilityInfo
         raise (f"Get entity(id: {entity.id}) neither ManufactureUnit nor ConsumerUnit")
 
 
-ALGO = "EOQ"
+ALGO = "BSP"
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/config.py:50]: revert

File: examples/supply_chain/rl/config.py [hunk 19]
```diff
@@ -64,10 +66,10 @@ def get_vlt_buffer_factor(entity: SupplyChainEntity, facility_info: FacilityInfo
 #     f"_{selection.value}"
 #     f"{'_storage_enlarged' if storage_enlarged else ''}"
 # )
-TOPOLOGY = "super_vendor"
+TOPOLOGY = "walmart_3_layers"
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/config.py:69]: revert

File: examples/supply_chain/rl/config.py [hunk 20]
```diff
@@ -64,10 +66,10 @@ def get_vlt_buffer_factor(entity: SupplyChainEntity, facility_info: FacilityInfo
...
 
-TRAIN_STEPS = 180
-EVAL_STEPS = 60
+TRAIN_STEPS = 1
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/config.py:71]: revert

File: examples/supply_chain/rl/forecastor/oracle_forecastor.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        sku_name = state["sku_name"]
+        facility_name = state["facility_name"]
+        future_start_offset = state["tick"] + 1
+        future_start_date = self.start_date + datetime.timedelta(future_start_offset)
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/oracle_forecastor.py:40]: `start_date` not used until the calling of this function

File: examples/supply_chain/rl/config.py (hunk 3 above)
Comment (Jinyu-W) [reply examples/supply_chain/rl/config.py:93]: can be removed due to the current "oracle_forecaster" implementation

File: examples/supply_chain/rl/forecastor/base_forecastor.py [hunk 22]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Microsoft Corporation.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecastor/base_forecastor.py:1]: `forecastor` -> `forecaster`

File: examples/supply_chain/rl/forecaster/base_forecaster.py [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+# Licensed under the MIT license.
+
+import pandas as pd
+
+
+class BaseForecaster(object):
+    def __init__(self, data_loader_conf: dict) -> None:
+        self.data_loader_conf = data_loader_conf
+
+    def forecast_future_demand(self, state: dict, df_history: pd.DataFrame) -> pd.DataFrame:
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/base_forecaster.py:11]: add `@abstractmethod` wrapper?

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+import os
+
+import pandas as pd
+
+from .base_forecaster import BaseForecaster
+
+
+class OracleForecaster(BaseForecaster):
+    """
+    The oracle forecaster load data from form files. Each form needs to include the following columns.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:14]: rename `form files` to `spreadsheets`

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+    FacilityName: facility name, should be defined in config.yml
+    SkuName: sku name, should be defined in config.yml
+    Date: date
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:17]: Is there any requirements to the date format?

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+    Price: sku's mean price in this facility
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:18]: what's the meaning?
price to selling to downstream (downstream facilities or end-customers)?
or sth. else?

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+    Demand: sku's demand in this facility.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:19]: not clear, similar to `Price`

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+    Cost: sku's mean cost in this facility.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:20]: not clear, similar to `Price`

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+                    df_list.append(pd.read_excel(oracle_file_path, parse_dates=["Date"]))
+                else:
+                    # TODO: Whether to terminate the program or just skip the file need to be decided.
+                    raise NotImplementedError
+        self.df_raws = pd.concat(df_list, axis=0)
+        self.df_raws.sort_values(by="Date", ascending=True)
+
+    def forecast_future_demand(self, state: dict, df_history: pd.DataFrame) -> pd.DataFrame:
+        """
+        To forecast future demand in oracle forecaster, following keys in state are needed.
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:47]:
```
The OracleForecaster filters data from the read-in file by the info in the given state. The required state keys include:
```

File: examples/supply_chain/rl/forecaster/oracle_forecaster.py [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+        sku_name: target sku name for forecasting.
+        facility_name: target facility name for forecasting
+        tick: date tick to indicate the days
+        start_date_time: start date in oracle file
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/forecaster/oracle_forecaster.py:51]: Is it the state date in oracle file, or the state date for the simulation?

File: examples/supply_chain/rl/env_sampler.py [hunk 31]
```diff
@@ -301,6 +305,17 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
         else:
             return 0.0
 
+    def _get_upstream_facility_id_to_sku_info_dict(self, facility_id: int, sku_id: int) -> Optional[Dict[int, SkuInfo]]:
+        # sku_id_to_facility_id_list: Dict[int, List[int]] = Dict[sku_id, upstream_facility_id_list]
+        sku_id_to_facility_id_list = self._facility_info_dict[facility_id].upstreams
+        if sku_id in sku_id_to_facility_id_list:
+            upstream_facility_id_list = sku_id_to_facility_id_list[sku_id]
+            facility_info_list = [self._facility_info_dict[facility_id] for facility_id in upstream_facility_id_list]
+            facility_id_to_sku_info = {facility.id: facility.skus[sku_id] for facility in facility_info_list}
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:314]: `for facility_info in facility_info_list`

File: examples/supply_chain/rl/env_sampler.py [hunk 32]
```diff
@@ -301,6 +305,17 @@ def _get_reward_for_entity(self, entity: SupplyChainEntity, bwt: Tuple[float, fl
         else:
             return 0.0
 
+    def _get_upstream_facility_id_to_sku_info_dict(self, facility_id: int, sku_id: int) -> Optional[Dict[int, SkuInfo]]:
+        # sku_id_to_facility_id_list: Dict[int, List[int]] = Dict[sku_id, upstream_facility_id_list]
+        sku_id_to_facility_id_list = self._facility_info_dict[facility_id].upstreams
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:310]: `sku_id_to_facility_id_list: Dict[int, List[int]] = self._facility_info_dict[facility_id].upstreams`

File: examples/supply_chain/rl/env_sampler.py [hunk 33]
```diff
@@ -379,6 +394,16 @@ def get_or_policy_state(self, entity: SupplyChainEntity) -> dict:
         if self._storage_capacity_dict is None:
             self._storage_capacity_dict = self._get_storage_capacity_dict_info()
 
+        upstream_facility_id_to_sku_info_dict = self._get_upstream_facility_id_to_sku_info_dict(
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:397]: add type hint

File: examples/supply_chain/rl/env_sampler.py [hunk 34]
```diff
@@ -198,6 +200,8 @@ def __init__(
 
         self.baseline_policy = ConsumerBaselinePolicy("baseline_eoq")
 
+        self.policy_action_by_quantity = [BaseStockPolicy]
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:203]: replace it by function?
```
def is_index_action(policy: xxx) -> bool:
    if isinstance(policy, BaseStockPolicy):
        return False
    return True
```

Comment (Jinyu-W) [reply examples/supply_chain/rl/env_sampler.py:203]: I found that there is no need for this list or function

File: examples/supply_chain/rl/env_sampler.py [hunk 35]
```diff
@@ -589,15 +618,17 @@ def _translate_to_env_action(
                 if isinstance(self._policy_dict[self._agent2policy[agent_id]], RLPolicy):
```
Comment (Jinyu-W) [inline examples/supply_chain/rl/env_sampler.py:618]:
```
if isinstance(self._policy_dict[self._agent2policy[agent_id]], RLPolicy):
    baseline_action = np.array(self._agent_state_dict[agent_id][-OR_NUM_CONSUMER_ACTIONS:])
    or_action = np.where(baseline_action == 1.0)[0][0]
    # action_idx = int(action[0] + or_action)
    action_idx = max(0, int(action[0] - 1 + or_action))
    product_unit_id: int = self._unit2product_unit[entity_id]
    action_quantity = int(
        int(action_idx) * max(1.0, self._cur_metrics["products"][product_unit_id]["demand_mean"]),
    )
else:
    action_quantity = action[0]
```

**Pull request merged**